File: sdks/javascript/src/resources/contacts.ts
/**
 * Contacts resource for managing marketing contacts
 */

import type { Mailat } from '../client'
import type {
  Contact,
  CreateContactOptions,
  UpdateContactOptions,
  ContactListOptions,
  PaginatedResponse,
  ContactList
} from '../types'

export class ContactsResource {
  constructor(private client: Mailat) {}

  /**
   * Create a new contact
   *
   * @example
   * ```typescript
   * const contact = await client.contacts.create({
   *   email: 'user@example.com',
   *   firstName: 'John',
   *   lastName: 'Doe',
   *   attributes: { company: 'Acme Inc' },
   *   tags: ['vip', 'early-adopter']
   * })
   * ```
   */
  async create(options: CreateContactOptions): Promise<Contact> {
    return this.client.post<Contact>('/contacts', options)
  }

  /**
   * Get contact by ID or email
   */
  async get(idOrEmail: string): Promise<Contact> {
    return this.client.get<Contact>(`/contacts/${encodeURIComponent(idOrEmail)}`)
  }

  /**
   * Update a contact
   */
  async update(id: string, options: UpdateContactOptions): Promise<Contact> {
    return this.client.put<Contact>(`/contacts/${id}`, options)
  }

  /**
   * Delete a contact
   */
  async delete(id: string): Promise<void> {
    await this.client.delete(`/contacts/${id}`)
  }

  /**
   * List contacts with pagination and filtering
   */
  async list(options: ContactListOptions = {}): Promise<PaginatedResponse<Contact>> {
    return this.client.get<PaginatedResponse<Contact>>('/contacts', {
      page: options.page,
      limit: options.limit,
      status: options.status,
      tag: options.tag,
      search: options.search
    })
  }

  /**
   * Search contacts
   */
  async search(query: string, options: { page?: number; limit?: number } = {}): Promise<Contact[]> {
    return this.client.get<Contact[]>('/contacts/search', {
      q: query,
      page: options.page,
      limit: options.limit
    })
  }

  /**
   * Import contacts in bulk
   *
   * @example
   * ```typescript
   * const result = await client.contacts.import([
   *   { email: 'user1@example.com', firstName: 'User 1' },
   *   { email: 'user2@example.com', firstName: 'User 2' }
   * ], { listId: 'newsletter-list' })
   * ```
   */
  async import(
    contacts: CreateContactOptions[],
    options: { listId?: string; tags?: string[] } = {}
  ): Promise<{ imported: number; failed: number; errors: { email: string; error: string }[] }> {
    return this.client.post('/contacts/import', {
      contacts,
      listId: options.listId,
      tags: options.tags
    })
  }

  /**
   * Unsubscribe a contact
   */
  async unsubscribe(email: string, options: { listId?: string } = {}): Promise<Contact> {
    return this.client.post<Contact>('/contacts/unsubscribe', {
      email,
      listId: options.listId
    })
  }

  /**
   * Add tags to a contact
   */
  async addTags(id: string, tags: string[]): Promise<Contact> {
    const contact = await this.get(id)
    const existingTags = contact.tags || []
    const newTags = [...new Set([...existingTags, ...tags])]
    return this.update(id, { tags: newTags })
  }

  /**
   * Remove tags from a contact
   */
  async removeTags(id: string, tags: string[]): Promise<Contact> {
    const contact = await this.get(id)
    const existingTags = contact.tags || []
    const newTags = existingTags.filter(t => !tags.includes(t))
    return this.update(id, { tags: newTags })
  }

  /**
   * Get lists a contact belongs to
   */
  async getLists(id: string): Promise<ContactList[]> {
    return this.client.get<ContactList[]>(`/contacts/${id}/lists`)
  }

  /**
   * Get contact activity/events
   */
  async getActivity(id: string, options: { page?: number; limit?: number } = {}): Promise<{
    events: {
      type: string
      timestamp: string
      data?: Record<string, unknown>
    }[]
    total: number
  }> {
    return this.client.get(`/contacts/${id}/activity`, {
      page: options.page,
      limit: options.limit
    })
  }
}

File: sdks/python/mailat/resources/contacts.py
"""Contacts resource for managing marketing contacts"""

from typing import TYPE_CHECKING, Any, Optional
from ..types import Contact, ContactList

if TYPE_CHECKING:
    from ..client import Mailat


class ContactsResource:
    """Resource for managing marketing contacts."""

    def __init__(self, client: "Mailat"):
        self._client = client

    def create(
        self,
        email: str,
        first_name: Optional[str] = None,
        last_name: Optional[str] = None,
        attributes: Optional[dict[str, Any]] = None,
        tags: Optional[list[str]] = None,
        list_ids: Optional[list[str]] = None,
    ) -> Contact:
        """Create a new contact."""
        payload = {"email": email}
        if first_name:
            payload["firstName"] = first_name
        if last_name:
            payload["lastName"] = last_name
        if attributes:
            payload["attributes"] = attributes
        if tags:
            payload["tags"] = tags
        if list_ids:
            payload["listIds"] = list_ids

        data = self._client.post("/contacts", json=payload)
        return Contact(**data)

    def get(self, id_or_email: str) -> Contact:
        """Get contact by ID or email."""
        data = self._client.get(f"/contacts/{id_or_email}")
        return Contact(**data)

    def update(
        self,
        contact_id: str,
        first_name: Optional[str] = None,
        last_name: Optional[str] = None,
        attributes: Optional[dict[str, Any]] = None,
        tags: Optional[list[str]] = None,
    ) -> Contact:
        """Update a contact."""
        payload = {}
        if first_name is not None:
            payload["firstName"] = first_name
        if last_name is not None:
            payload["lastName"] = last_name
        if attributes is not None:
            payload["attributes"] = attributes
        if tags is not None:
            payload["tags"] = tags

        data = self._client.put(f"/contacts/{contact_id}", json=payload)
        return Contact(**data)

    def delete(self, contact_id: str) -> None:
        """Delete a contact."""
        self._client.delete(f"/contacts/{contact_id}")

    def list(
        self,
        page: int = 1,
        limit: int = 50,
        status: Optional[str] = None,
        tag: Optional[str] = None,
        search: Optional[str] = None,
    ) -> dict[str, Any]:
        """List contacts with pagination."""
        params = {"page": page, "limit": limit}
        if status:
            params["status"] = status
        if tag:
            params["tag"] = tag
        if search:
            params["search"] = search

        return self._client.get("/contacts", params=params)

    def search(self, query: str, page: int = 1, limit: int = 50) -> list[Contact]:
        """Search contacts."""
        data = self._client.get("/contacts/search", params={"q": query, "page": page, "limit": limit})
        return [Contact(**c) for c in data]

    def import_contacts(
        self,
        contacts: list[dict[str, Any]],
        list_id: Optional[str] = None,
        tags: Optional[list[str]] = None,
    ) -> dict[str, Any]:
        """Import contacts in bulk."""
        payload = {"contacts": contacts}
        if list_id:
            payload["listId"] = list_id
        if tags:
            payload["tags"] = tags

        return self._client.post("/contacts/import", json=payload)

    def unsubscribe(self, email: str, list_id: Optional[str] = None) -> Contact:
        """Unsubscribe a contact."""
        payload = {"email": email}
        if list_id:
            payload["listId"] = list_id

        data = self._client.post("/contacts/unsubscribe", json=payload)
        return Contact(**data)

    def add_tags(self, contact_id: str, tags: list[str]) -> Contact:
        """Add tags to a contact."""
        contact = self.get(contact_id)
        existing_tags = contact.tags or []
        new_tags = list(set(existing_tags + tags))
        return self.update(contact_id, tags=new_tags)

    def remove_tags(self, contact_id: str, tags: list[str]) -> Contact:
        """Remove tags from a contact."""
        contact = self.get(contact_id)
        existing_tags = contact.tags or []
        new_tags = [t for t in existing_tags if t not in tags]
        return self.update(contact_id, tags=new_tags)

    def get_lists(self, contact_id: str) -> list[ContactList]:
        """Get lists a contact belongs to."""
        data = self._client.get(f"/contacts/{contact_id}/lists")
        return [ContactList(**l) for l in data]
